File: Exp 2 Searching Algorithms /binary_search.cpp
#include <iostream>
using namespace std;


// Function to insert the array elements
void insert_sorted_array_elements(int *arr, int size)
{
    cout << "Enter the elements of the array" << endl;
    for (int i = 0; i < size; i++)
    {
        cout << "- Element " << i << ": ";
        cin >> arr[i];
        if(i != 0 && arr[i - 1] > arr[i]){
            cout << "Array should be in a sorted form!" << endl;
            break;
        }
    }
}

// Function to display the array
void display_array(int *arr, int size)
{
    for (int i= 0; i < size; i++) cout << arr[i] << " ";
    cout << endl;
}

// function to binary search
int binary_search(int arr[], int left, int right, int target)
{
    int mid = left + (right - left) / 2;
    if(arr[mid] == target) return mid;
    else if(arr[mid] < target) return binary_search(arr, mid+1, right, target);
    else return binary_search(arr, left, mid-1, target);
}

int main(){
    // Creating the array
    int size;
    cout << "Enter the size of the array: ";
    cin >> size;

    // Insert the elements in the array
    int arr[size];
    insert_sorted_array_elements(arr, size);

    int target;
    cout << "Enter the element to be searched in the array: ";
    cin >> target;

    // Display the created array
    cout << endl << "Your array: ";
    display_array(arr,size);

    // Binary search the target
    cout << "Binary Searching ..." << endl;
    cout << "=> Element found at index " << binary_search(arr, 0, size, target) << endl;

    return 0;
}

File: Exp 2 Searching Algorithms /binary_search.py
def insert_sorted_array_elements(size):
    """Function to insert the array elements"""
    print("Enter the elements of the array")
    arr = []
    for i in range(size):
        element = int(input(f"- Element {i}: "))
        if i != 0 and arr[i - 1] > element:
            print("Array should be in a sorted form!")
            break
        arr.append(element)
    return arr


def display_array(arr):
    """Function to display the array"""
    print(" ".join(map(str, arr)))


def binary_search(arr, left, right, target):
    """Function to binary search"""
    mid = left + (right - left) // 2
    if arr[mid] == target:
        return mid
    elif arr[mid] < target:
        return binary_search(arr, mid + 1, right, target)
    else:
        return binary_search(arr, left, mid - 1, target)


def main():
    # Creating the array
    size = int(input("Enter the size of the array: "))
    
    # Insert the elements in the array
    arr = insert_sorted_array_elements(size)
    
    target = int(input("Enter the element to be searched in the array: "))
    
    # Display the created array
    print("\nYour array: ", end="")
    display_array(arr)
    
    # Binary search the target
    print("Binary Searching ...")
    result = binary_search(arr, 0, size - 1, target)
    print(f"=> Element found at index {result}")


if __name__ == "__main__":
    main()
